### PR TITLE
HPCC-25531 Roxie decryption should be moved off socket reader thread

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/udplib/udptrs.cpp [hunk 1]
```diff
@@ -204,7 +204,15 @@ class UdpReceiverEntry : public IUdpReceiverEntry
...
-                    aesEncrypt(key, sizeof(key), buffer->data, length, encryptBuffer.clear());
+                    encryptBuffer.clear();
+                    encryptBuffer.append(sizeof(UdpPacketHeader), header);    // We don't encrypt the header
+                    length -= sizeof(UdpPacketHeader);
+                    const char *data = buffer->data + sizeof(UdpPacketHeader);
+                    aesEncrypt(key, sizeof(key), data, length, encryptBuffer);
+                    header->length = encryptBuffer.length();
+                    encryptBuffer.writeDirect(0, sizeof(UdpPacketHeader), header);   // Only really need length updating
+                    assert(length <= DATA_PAYLOAD);
+                    DBGLOG("ENCRYPT: Writing %u bytes to data socket", encryptBuffer.length());
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:215]: Excessive tracing...

File: roxie/udplib/udptrs.cpp [hunk 2]
```diff
@@ -204,7 +204,15 @@ class UdpReceiverEntry : public IUdpReceiverEntry
             {
                 if (encrypted)
                 {
-                    aesEncrypt(key, sizeof(key), buffer->data, length, encryptBuffer.clear());
+                    encryptBuffer.clear();
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:207]: could ensure capacity of DATA_PAYLOAD to avoid reallocation.

File: roxie/udplib/udptrs.cpp [hunk 3]
```diff
@@ -704,10 +713,11 @@ class CSendManager : implements ISendManager, public CInterface
 public:
     IMPLEMENT_IINTERFACE;
 
-    CSendManager(int server_flow_port, int data_port, int client_flow_port, int sniffer_port, const IpAddress &sniffer_multicast_ip, int q_size, int _numQueues, const IpAddress &_myIP, TokenBucket *_bucket, bool encrypted)
+    CSendManager(int server_flow_port, int data_port, int client_flow_port, int sniffer_port, const IpAddress &sniffer_multicast_ip, int q_size, int _numQueues, const IpAddress &_myIP, TokenBucket *_bucket, bool _encrypted)
         : bucket(_bucket),
           myIP(_myIP),
-          receiversTable([_myIP, _numQueues, q_size, server_flow_port, data_port, encrypted](const ServerIdentifier &ip) { return new UdpReceiverEntry(ip.getIpAddress(), _myIP, _numQueues, q_size, server_flow_port, data_port, encrypted);})
+          receiversTable([_myIP, _numQueues, q_size, server_flow_port, data_port, _encrypted](const ServerIdentifier &ip) { return new UdpReceiverEntry(ip.getIpAddress(), _myIP, _numQueues, q_size, server_flow_port, data_port, _encrypted);}),
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:719]: will need rebasing

**Review by ghalliday — COMMENTED** [review]
Generally looks good

File: roxie/udplib/udptrs.cpp (hunk 1 above)
Comment (mckellyln) [reply roxie/udplib/udptrs.cpp:215]: This looks good, but I am missing where the extra possible 16 bytes of encryptBuffer are allocated ?

Comment (mckellyln) [reply roxie/udplib/udptrs.cpp:215]: ok, I see inside MemoryBuffer &aesEncrypt(const void *key, size_t keylen, const void *input, size_t inlen, MemoryBuffer &output) is:
output.reserveTruncate(truncInLen + 16));

**Review by mckellyln — DISMISSED** [review]
Approved

**Comment by ghalliday** [comment]
@richardkchapman do you want to make this non draft and target 8.0.x?

**Comment by richardkchapman** [comment]
Yes, but I want the udpretry merged first so I only have to rebase once

File: roxie/udplib/udptrs.cpp (hunk 2 above)
Comment (richardkchapman) [reply roxie/udplib/udptrs.cpp:423]: Or length+16

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Review by ghalliday — APPROVED** [review]

**Pull request merged**